src/measure_theory/function/ae_eq_of_integral.lean

```diff
@@ -0,0 +1,486 @@
+
+Generally useful lemmas which are not related to integrals:
+* `ae_eq_zero_of_forall_inner`: if for all constants `c`, `λ x, inner c (f x) =ᵐ[μ] 0` then
+  `f =ᵐ[μ] 0`.
+* `ae_eq_zero_of_forall_dual`: if for all constants `c` in the dual space, `λ x, c (f x) =ᵐ[μ] 0`
+  then `f =ᵐ[μ] 0`.
+
+## TODO(s)
+
+* Extend the results from a Hilbert space to a Banach space. Most lemmas about inner products
+  should be equally valid about the product with an element of the dual and
+  `ae_eq_zero_of_forall_dual` should be used instead of `ae_eq_zero_of_forall_inner`.
+
+-/
+
+open measure_theory topological_space normed_space filter
+
+open_locale ennreal measure_theory
+
+namespace measure_theory
+
+
+section ae_eq_of_forall
+
+variables {α E 𝕜 : Type*} {m : measurable_space α} {μ : measure α} [is_R_or_C 𝕜]
+
+lemma ae_eq_zero_of_forall_inner [inner_product_space 𝕜 E] [second_countable_topology E]
+  {f : α → E} (hf : ∀ c : E, (λ x, (inner c (f x) : 𝕜)) =ᵐ[μ] 0) :
+  f =ᵐ[μ] 0 :=
+begin
+  let s := dense_seq E,
+  have hs : dense_range s := dense_range_dense_seq E,
+  have hf' : ∀ᵐ x ∂μ, ∀ n : ℕ, inner (s n) (f x) = (0 : 𝕜), from ae_all_iff.mpr (λ n, hf (s n)),
+  refine hf'.mono (λ x hx, _),
+  rw [pi.zero_apply, ← inner_self_eq_zero],
+  have h_closed : is_closed {c : E | inner c (f x) = (0 : 𝕜)},
+    from is_closed_eq (continuous_id.inner continuous_const) continuous_const,
+  exact @is_closed_property ℕ E _ s (λ c, inner c (f x) = (0 : 𝕜)) hs h_closed (λ n, hx n) _,
+end
+
+local notation `⟪`x`, `y`⟫` := y x
+
+variables (𝕜)
+
+lemma ae_eq_zero_of_forall_dual [normed_group E] [normed_space 𝕜 E]
+  [second_countable_topology (dual 𝕜 E)]
+  {f : α → E} (hf : ∀ c : dual 𝕜 E, (λ x, ⟪f x, c⟫) =ᵐ[μ] 0) :
+  f =ᵐ[μ] 0 :=
+begin
+  let s := dense_seq (dual 𝕜 E),
+  have hs : dense_range s := dense_range_dense_seq _,
+  have hfs : ∀ n : ℕ, ∀ᵐ x ∂μ, ⟪f x, s n⟫ = (0 : 𝕜), from λ n, hf (s n),
+  have hf' : ∀ᵐ x ∂μ, ∀ n : ℕ, ⟪f x, s n⟫ = (0 : 𝕜), by rwa ae_all_iff,
+  refine hf'.mono (λ x hx, eq_zero_of_forall_dual_eq_zero 𝕜 (λ c, _)),
+  have h_closed : is_closed {c : dual 𝕜 E | ⟪f x, c⟫ = (0 : 𝕜)},
+  { refine is_closed_eq _ continuous_const,
+    have h_fun_eq : (λ (c : dual 𝕜 E), ⟪f x, c⟫) = inclusion_in_double_dual 𝕜 E (f x),
+      by { ext1 c, rw ← dual_def 𝕜 E (f x) c, },
+    rw h_fun_eq,
+    continuity, },
+  exact @is_closed_property ℕ (dual 𝕜 E) _ s (λ c, ⟪f x, c⟫ = (0 : 𝕜)) hs h_closed (λ n, hx n) c,
+end
+
+variables {𝕜}
+
+end ae_eq_of_forall
+
+
+variables {α 𝕜 E : Type*} [is_R_or_C 𝕜] [measurable_space 𝕜] [borel_space 𝕜]
+  {m m0 : measurable_space α} {μ : measure α} {s t : set α}
+  [inner_product_space 𝕜 E] [measurable_space E] [borel_space E] [second_countable_topology E]
+  [complete_space E] [normed_space ℝ E]
+  {p : ℝ≥0∞}
+
+section ae_eq_of_forall_set_integral_eq
+
+lemma ae_const_le_iff_forall_lt_measure_zero (f : α → ℝ) (c : ℝ) :
+  (∀ᵐ x ∂μ, c ≤ f x) ↔ ∀ b < c, μ {x | f x ≤ b} = 0 :=
```

Comment:
this would also apply to ennreal or nnreal (and I can imagine this could be useful). Can you come up with a typeclass that covers these all? (linearly ordered space with order topology and first countable topology, maybe?)

Reply:
This was not as trivial as I thought it would be, because of the different ways that rationals are converted to reals, nnreal and ennreal. I could get three lemmas applying to the different cases, with a single auxiliary lemma providing the proof. Please tell me if you see a simpler way.

Reply:
I have pushed a proof that covers all situations in one go, replacing the rationals by an abstract sequence. But I'm not completely satisfied about it as this does not cover the case of the integers, while the result is still true there. I'll try to come up with an even more general version.

Reply:
Done.